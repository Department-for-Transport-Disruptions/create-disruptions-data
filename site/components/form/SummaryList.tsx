import Link from "next/link";
import { ReactNode } from "react";

interface SummaryListProps {
    rows: {
        header?: string;
        value: string | ReactNode;
        actions?: {
            link:
                | string
                | {
                      pathname: string;
                      query: { template?: string | undefined; return: string } | { template: string } | null;
                  };
            actionName: string;
        }[];
    }[];
}

const SummaryList = ({ rows }: SummaryListProps) => (
    <dl className="govuk-summary-list">
        {rows.map((row) => (
            <div className="govuk-summary-list__row" key={row.header}>
                <dt className="govuk-summary-list__key">{row.header}</dt>
                <dd className="govuk-summary-list__value">{row.value}</dd>
                <dd className="govuk-summary-list__actions">
                    <ul className="govuk-summary-list__actions-list">
                        {row.actions && row.actions.length > 0
                            ? row.actions.map((action) => (
                                  <li className="govuk-summary-list__actions-list-item" key={row.header}>
                                      <Link className="govuk-link" href={action.link}>
                                          {action.actionName}
                                          <span className="govuk-visually-hidden">{row.header}</span>
                                      </Link>
                                  </li>
                              ))
                            : null}
                    </ul>
                </dd>
            </div>
        ))}
    </dl>
);

export default SummaryList;
