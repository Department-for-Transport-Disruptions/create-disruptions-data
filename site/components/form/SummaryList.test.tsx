import renderer from "react-test-renderer";
import SummaryList from "./SummaryList";
import { describe, expect, it } from "vitest";

describe("SummaryList Component", () => {
    it("matches the snapshot with rows, headers, values, and actions", () => {
        const rows = [
            {
                header: "Header 1",
                value: "Value 1",
                actions: [
                    {
                        link: "/path1",
                        actionName: "Action 1",
                    },
                ],
            },
            {
                header: "Header 2",
                value: "Value 2",
                actions: [
                    {
                        link: "/path2",
                        actionName: "Action 2",
                    },
                ],
            },
        ];

        const tree = renderer.create(<SummaryList rows={rows} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("matches the snapshot with rows without actions", () => {
        const rows = [
            {
                header: "Header 1",
                value: "Value 1",
            },
        ];

        const tree = renderer.create(<SummaryList rows={rows} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("matches the snapshot with ReactNode as value", () => {
        const rows = [
            {
                header: "Header 1",
                value: <span>ReactNode Value</span>,
            },
        ];

        const tree = renderer.create(<SummaryList rows={rows} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders the correct href and text for actions", () => {
        const rows = [
            {
                header: "Header 1",
                value: "Value 1",
                actions: [
                    {
                        link: "/path1",
                        actionName: "Action 1",
                    },
                ],
            },
        ];

        const tree = renderer.create(<SummaryList rows={rows} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders visually-hidden text for each action", () => {
        const rows = [
            {
                header: "Header 1",
                value: "Value 1",
                actions: [
                    {
                        link: "/path1",
                        actionName: "Action 1",
                    },
                ],
            },
        ];

        const tree = renderer.create(<SummaryList rows={rows} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
