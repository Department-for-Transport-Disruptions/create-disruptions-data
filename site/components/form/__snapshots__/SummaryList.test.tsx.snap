// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SummaryList Component > matches the snapshot with ReactNode as value 1`] = `
<dl
  className="govuk-summary-list"
>
  <div
    className="govuk-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key"
    >
      Header 1
    </dt>
    <dd
      className="govuk-summary-list__value"
    >
      <span>
        ReactNode Value
      </span>
    </dd>
    <dd
      className="govuk-summary-list__actions"
    >
      <ul
        className="govuk-summary-list__actions-list"
      />
    </dd>
  </div>
</dl>
`;

exports[`SummaryList Component > matches the snapshot with rows without actions 1`] = `
<dl
  className="govuk-summary-list"
>
  <div
    className="govuk-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key"
    >
      Header 1
    </dt>
    <dd
      className="govuk-summary-list__value"
    >
      Value 1
    </dd>
    <dd
      className="govuk-summary-list__actions"
    >
      <ul
        className="govuk-summary-list__actions-list"
      />
    </dd>
  </div>
</dl>
`;

exports[`SummaryList Component > matches the snapshot with rows, headers, values, and actions 1`] = `
<dl
  className="govuk-summary-list"
>
  <div
    className="govuk-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key"
    >
      Header 1
    </dt>
    <dd
      className="govuk-summary-list__value"
    >
      Value 1
    </dd>
    <dd
      className="govuk-summary-list__actions"
    >
      <ul
        className="govuk-summary-list__actions-list"
      >
        <li
          className="govuk-summary-list__actions-list-item"
        >
          <a
            className="govuk-link"
            href="/path1"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            Action 1
            <span
              className="govuk-visually-hidden"
            >
              Header 1
            </span>
          </a>
        </li>
      </ul>
    </dd>
  </div>
  <div
    className="govuk-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key"
    >
      Header 2
    </dt>
    <dd
      className="govuk-summary-list__value"
    >
      Value 2
    </dd>
    <dd
      className="govuk-summary-list__actions"
    >
      <ul
        className="govuk-summary-list__actions-list"
      >
        <li
          className="govuk-summary-list__actions-list-item"
        >
          <a
            className="govuk-link"
            href="/path2"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            Action 2
            <span
              className="govuk-visually-hidden"
            >
              Header 2
            </span>
          </a>
        </li>
      </ul>
    </dd>
  </div>
</dl>
`;

exports[`SummaryList Component > renders the correct href and text for actions 1`] = `
<dl
  className="govuk-summary-list"
>
  <div
    className="govuk-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key"
    >
      Header 1
    </dt>
    <dd
      className="govuk-summary-list__value"
    >
      Value 1
    </dd>
    <dd
      className="govuk-summary-list__actions"
    >
      <ul
        className="govuk-summary-list__actions-list"
      >
        <li
          className="govuk-summary-list__actions-list-item"
        >
          <a
            className="govuk-link"
            href="/path1"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            Action 1
            <span
              className="govuk-visually-hidden"
            >
              Header 1
            </span>
          </a>
        </li>
      </ul>
    </dd>
  </div>
</dl>
`;

exports[`SummaryList Component > renders visually-hidden text for each action 1`] = `
<dl
  className="govuk-summary-list"
>
  <div
    className="govuk-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key"
    >
      Header 1
    </dt>
    <dd
      className="govuk-summary-list__value"
    >
      Value 1
    </dd>
    <dd
      className="govuk-summary-list__actions"
    >
      <ul
        className="govuk-summary-list__actions-list"
      >
        <li
          className="govuk-summary-list__actions-list-item"
        >
          <a
            className="govuk-link"
            href="/path1"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            Action 1
            <span
              className="govuk-visually-hidden"
            >
              Header 1
            </span>
          </a>
        </li>
      </ul>
    </dd>
  </div>
</dl>
`;
